### PR TITLE
[#44945] Implementing `WorkPackage.at_timestamp`

### PR DESCRIPTION
**Pull request by fiedl**
In the context of the [baseline-comparison feature](https://community.openproject.org/projects/openproject/work_packages/26448), this pull request adds convenience methods to retrieve historic data of journable objects like work packages.

As the convenience methods introduced in this pull request are not used, this pull request does not change behaviour of the application, yet.

### Context

Component_Overview-3

### Retrieve historic state of one journable

```ruby
WorkPackage.find(1).at_timestamp(1.year.ago)
```

This historic state of the work package is still a `WorkPackage` object, but it contains data from the `work_package_journals` table. It knows that it represents historic data and, therefore, cannot be modified.

```ruby
work_package = WorkPackage.find(1)
historic_work_package = work_package.at_timestamp(1.year.ago)

historic_work_package.id         # => 1  (same as work_package.id)
historic_work_package.historic?  # => true
historic_work_package.save       # raises `ActiveRecord::ReadOnlyRecord`
```

### Rollback to historic state of one journable

```ruby
WorkPackage.find(1).at_timestamp(1.year.ago).rollback!
```

This updates the work-package record with the attributes from the historic record, which does not remove journal entries, but, in fact, creates a new journal entry.

```ruby
WorkPackage.find(1).update! WorkPackage.find(1).at_timestamp(1.year.ago).attributes
```

### Retrieve historic state of multiple journables

```ruby
WorkPackage.at_timestamp(1.year.ago)
```

This returns an `ActiveRecord::Relation` that collects `WorkPackage` objects, which contain data from the `work_package_journals` table. These objects know that they represent historic data and, therefore, cannot be modified.

### Filter on historic data

```ruby
WorkPackage.at_timestamp(1.year.ago).where(assigned_to_id: 1)

# or, equivalently:
WorkPackage.where(assigned_to_id: 1).at_timestamp(1.year.ago)
```

This returns an `ActiveRecord::Relation` that represents all work packages that have been assigned to `User.find(1)` one year ago, regardless of whom they are assigned now.

### Plucking on historic data

```ruby
WorkPackage.at_timestamp(1.year.ago).where(assigned_to_id: 1).pluck :id
```

This returns the IDs of all `WorkPackage` records that have been assigned to `User.find(1)` one year ago.

These IDs are the IDs of the work-package records, not the work-package-journal records. Therefore, when fetching the work-package records from the database using these IDs, this will fetch the records in their current state, not the historic state.

```ruby
work_package_ids = WorkPackage.at_timestamp(1.year.ago).where(assigned_to_id: 1).pluck :id
work_packages = WorkPackage.where(id: work_package_ids)

work_packages.first.historic?  # => false
```

### Sub queries

```ruby
sub_query = WorkPackage.at_timestamp(1.year.ago).where(assigned_to_id: 1)
WorkPackage.where(id: sub_query)
```

This returns the IDs of all `WorkPackage` records that have been assigned to `User.find(1)` one year ago. As this is using a sub query, which is an `ActiveRecord::Relation`, this only hits the database once.

### Advanced queries

```ruby
WorkPackage.at_timestamp(1.year.ago).maximum(:estimated_hours)
```

This finds largest `estimated_hours` considering all work packages in their states one year ago.

```ruby
WorkPackage.at_timestamp(1.year.ago).group(:assigned_to_id).count  # => {1 => 15, ...}
```

This counts how many work packages have been assigned to each user one year ago.

```ruby
WorkPackage.at_timestamp(1.year.ago).group(:assigned_to_id).having('estimated_hours > 3').count  # => {1 => 12, ...}
```

This counts how many work packages have been assigned to each user one year ago, but only considers work packages with hour estimates over 3 hours.

#### Filtering on associated tables

When adding where conditions on associated tables, one might expect to also filter on historic data.

```ruby
# Does not work:
WorkPackage.at_timestamp(1.year.ago).joins(:time_entries).where(time_entries: {hours: 1.0})
```

Because `WorkPackage.at_timestamp` only searches the journal of the work packages, not the journal of the time entries, this does **not** return work packages that had time entries of 1 hour one year ago, but instead returns the state of the work packages of one year ago that do now have time entries of 1 hour.

#### Manual selects

Watch out when using the `id` column in manually specified parts of the query.

The wrapper introduced with `at_timestamp` internally modifies the active-record relation to query the work-package-journals table rather than the work-packages table. So, for example, when running

```ruby
WorkPackage.at_timestamp(1.day.ago).select("work_pacakges.id")
# raises `ActiveRecord::StatementInvalid: PG::UndefinedTable`
```

you would get an error because the `work_packges` table is not know to that query.

As a workaround, you may use the column `journables.id`, which is induced by the wrapper:

```ruby
WorkPackage.at_timestamp(1.day.ago).select("journables.id")
```

### Baseline comparison

Consider this pseudo API call to `GET /api/v3/work_packages` with the following parameters:

- `author_id: 1` specifies the author user of the work packges
- `offset: 2` specifies the page number of the paginated result
- `sortBy: [["status", "asc"]]` specifies the sort order to the result
- `timestamps: ["2021-09-04T13:21:14.452+00:00", "2022-09-04T13:21:14.452+00:00"]` specifies the timestamps at which historic data should be searched

The result should contain

- the work packages that match the filter `author_id` at **any** of the given `timestamps`
- the work-package data at **all** given `timestamps` in order for the user interface to show what has changed comparing the data at those `timestamps`.

Using the convenience methods introduced by this pull request, this is still a several-steps process:

```ruby
# given: params[:author_id], params[:offset], params[:sort_by], params[:timestamps]

# Apply filters
work_packages_relation = WorkPackage \
    .where(author_id: params[:author_id])

# Get IDs of work packages matching the filters at any of the given timestamps.
work_packages_for_all_timestamps = params[:timestamps].collect do |timestamp|
  work_packages_relation.at_timestamp(timestamp)
end.reduce(:or)

# Order the result
ordered_work_packages_relation = WorkPackage \
    .where(id: work_packages_for_all_timestamps) \
    .order(params[:sort_by])

# Paginate the result
paginated_work_packages_relation = \
    ordered_work_packages_relation \
    .paginate(page: params[:offset])
    
# Of all the matching work packages on that page, retrieve the historic attributes
# at all of the given timestamps.
paginated_work_packages_by_timestamp = {}
params[:timestamps].each do |timestamp|
  paginated_work_packages_by_timestamp[timestamp] = WorkPackage \
      .where(id: paginated_work_packages_relation) \
      .at_timestamp(timestamp)
end
```

### See also

- **Implementation Overview**: https://github.com/fiedl/openproject/issues/4
- **Implementation Tasks**: https://community.openproject.org/projects/openproject/work_packages/43483, https://community.openproject.org/projects/openproject/work_packages/44945
- **Epic**: https://community.openproject.org/projects/openproject/work_packages/26448
- https://github.com/fiedl/openproject/issues/3

### CODE REVIEW

**Comment by fiedl** [comment]
@ulferts Ready for review.

This first pull request adds convenience methods, but does not call them. Therefore, this pull request does not change the behaviour of the application.

File: app/models/journable/historic_active_record_relation.rb [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+  end
+
+  # Add a timestamp condition: Select the work package journals that are the
+  # current ones at the given timestamp.
+  #
+  def add_timestamp_condition(relation)
+    relation \
+        .joins("INNER JOIN \"journals\" ON \"journals\".\"data_type\" = '#{model.journal_class.name}'
+            AND \"journals\".\"data_id\" = \"#{model.journal_class.table_name}\".\"id\"".gsub("\n", "")) \
+        .merge(Journal.at_timestamp(timestamp))
```
Comment (ulferts) [inline app/models/journable/historic_active_record_relation.rb]: There is no alias on this which might lead to conflicts.

File: app/models/journable/historic_active_record_relation.rb [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+         Arel::Nodes::LessThanOrEqual,
+         Arel::Nodes::GreaterThan,
+         Arel::Nodes::GreaterThanOrEqual
+      if predicate.left.relation == arel_table or predicate.left.relation == journal_class.arel_table
+        case predicate.left.name
+        when "id"
+          predicate.left.name = "journable_id"
+          predicate.left.relation = Journal.arel_table
+        when "updated_at"
+          predicate.left.name = "created_at"
```
Comment (ulferts) [inline app/models/journable/historic_active_record_relation.rb]: This should be kept as `updated_at`. We already talked about this so this is more a reminder.

File: app/models/journal/timestamps.rb [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    # a journable at a given timestamp.
+    #
+    def at_timestamp(timestamp)
+      unless timestamp.kind_of? ActiveSupport::TimeWithZone or timestamp.kind_of? DateTime
+        raise ArgumentError, "Expected timestamp to be an ActiveSupport::TimeWithZone or DateTime"
+      end
+
+      timestamp = timestamp.in_time_zone if timestamp.kind_of? DateTime
+
+      where(id: where(created_at: (..timestamp)) \
```
Comment (ulferts) [inline app/models/journal/timestamps.rb:72]: Adding a limitation on the `journable_type` will reduce the number of journals to be processed. Unfortunately not by that much since 'WorkPackage' journals tend to be in the vast majority.

With it being added to the WHERE clause, it could be removed from both the ORDER BY as well as from the DISTINCT ON statement. However, tests on production data indicate that having both the added WHERE condition as well as keeping the column in the `distinct_on` leads to the best performance.

File: app/models/journable/historic_active_record_relation.rb [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+  #
+  # Also, add the `timestamp` as column so that we have it as attribute in our model.
+  #
+  def select_columns_from_the_appropriate_tables(relation)
+    if relation.select_values.count == 0
+      relation = relation.select("'#{timestamp}' as timestamp,
+          #{model.journal_class.table_name}.*,
+          journals.journable_id as id,
+          journables.created_at as created_at,
+          journals.created_at as updated_at".gsub("\n", ""))
```
Comment (ulferts) [inline app/models/journable/historic_active_record_relation.rb]: We already talked about  this but as a reminder: `journals.updated_at` should be used here instead.

Comment (ulferts) [reply app/models/journable/historic_active_record_relation.rb]: Can we change `.gsub("\n", "")` to `.gsub(/\n\s+/, " ")`, or switch to a heredoc with `squish`? Admittedly it is nitpicking but it does read better in the console.

File: app/models/journable/historic_active_record_relation.rb [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+  private
+
+  # Switch the database table, e.g. from `work_packages` to `work_package_journals`.
+  #
+  def switch_to_journals_database_table(relation)
+    relation.instance_variable_set :@table, model.journal_class.arel_table
+    relation
+  end
+
+  # Modify there where clauses such that e.g. the work-packages table is substituted
```
Comment (ulferts) [inline app/models/journable/historic_active_record_relation.rb]: Typo: `there` -> `the`

File: app/models/journable/timestamps.rb [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+  def at_timestamp(timestamp)
+    if journal = journals.at_timestamp(timestamp).first
+      attributes = journal.data.attributes.merge(
+        {
+          "id" => id,
+          "created_at" => created_at,
+          "updated_at" => journal.updated_at,
+          "timestamp" => timestamp,
+          "position" => (position if respond_to? :position)
+        }
```
Comment (ulferts) [inline app/models/journable/timestamps.rb:99]: The `lock_version` property is not included which leads to failures when doing:
` WorkPackage.first.at_timestamp(1.year.ago).lock_version`. The property could be mapped to the `version` column of the `journals` table.

File: app/models/journable/historic_active_record_relation.rb [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+  # would return the `id` of the journal table rather than the `id` of the
+  # journable table, which would be expected from the syntax:
+  #
+  #     WorkPackage.where(assigned_to_id: 123).at_timestamp(1.year.ago).pluck(:id)
+  #
+  def pluck(*column_names)
+    column_names.map! { |column_name| column_name == :id ? 'journals.journable_id' : column_name }
+    arel
+    super
+  end
```
Comment (ulferts) [inline app/models/journable/historic_active_record_relation.rb:97]: There are more properties for which the behaviour might be surprising but those will fail hard:
* created_at
* updated_at
* lock_version
* position

From the top of my head, I don't think they are used within the codebase so it might be acceptable to let them simply fail hard until they are employed. On the other hand, it is a recurring pattern to have to map the properties so if some map could be established that is used in the various cases, further errors can be avoided. At least `created_at` and `updated_at` are non work package specific.

File: app/models/journable/historic_active_record_relation.rb [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+        .merge(Journal.at_timestamp(timestamp))
+  end
+
+  # Join the journables table itself because we need to take the `created_at` attribute from that.
+  # The `created_at` column is not present in the `work_package_journals` table.
+  #
+  def add_join_on_journables_table_with_created_at_column(relation)
+    relation \
+        .joins("INNER JOIN (SELECT id, created_at FROM \"#{model.table_name}\") AS journables
+            ON \"journables\".\"id\" = \"journals\".\"journable_id\"".gsub("\n", ""))
```
Comment (ulferts) [inline app/models/journable/historic_active_record_relation.rb]: The `position` needs to also be included from here.

File: app/models/journable/historic_active_record_relation.rb [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+        join.kind_of?(Arel::Nodes::StringJoin) and
+        join.left.include?("INNER JOIN \"journals\" ON \"journals\".\"data_type\"")
+      end
+      if journals_join_index
+        journals_join = array_of_joins[journals_join_index]
+        array_of_joins.delete_at(journals_join_index)
+        array_of_joins.insert(0, journals_join)
+      end
+    end
+    arel
```
Comment (ulferts) [inline app/models/journable/historic_active_record_relation.rb]: Instead of moving the join up afterwards, wouldn't it be easier to place the join at the first position in the `add_timestamp_condition` directly?

File: app/models/journable/historic_active_record_relation.rb [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+  end
+
+  # Replace table names in sql strings, e.g.
+  #
+  #     "work_package.id" => "journals.journable_id"
+  #     "work_package.subject" => "work_package_journals.subject"
+  #
+  def gsub_table_names_in_sql_string!(sql_string)
+    sql_string.gsub! /(?<!_)#{model.table_name}\.updated_at/, "journals.created_at"
+    sql_string.gsub! "\"#{model.table_name}\".\"updated_at\"", "\"journals\".\"created_at\""
```
Comment (ulferts) [inline app/models/journable/historic_active_record_relation.rb]: Should both be switched to `journals.updated_at`

File: app/models/journable/historic_active_record_relation.rb (hunk 9 above)
Comment (ulferts) [reply app/models/journable/historic_active_record_relation.rb]: If not, I think it worthwhile to move the SQL string in `add_timestamp_condition` into its own method so that it can be called to compare the joins to.

File: app/models/journable/historic_active_record_relation.rb [hunk 11]
```diff
@@ -0,0 +1,290 @@
...
+    sql_string.gsub! /(?<!_)#{model.table_name}\.created_at/, "journables.created_at"
+    sql_string.gsub! "\"#{model.table_name}\".\"created_at\"", "\"journables\".\"created_at\""
+    sql_string.gsub! /(?<!_)#{model.table_name}\.id/, "journals.journable_id"
+    sql_string.gsub! "\"#{model.table_name}\".\"id\"", "\"journals\".\"journable_id\""
+    sql_string.gsub! /(?<!_)#{model.table_name}\./, "#{model.journal_class.table_name}."
+    sql_string.gsub! "\"#{model.table_name}\".", "\"#{model.journal_class.table_name}\"."
```
Comment (ulferts) [inline app/models/journable/historic_active_record_relation.rb:313]: Again, `lock_version` and `position` need to be mapped.

**Review by ulferts — CHANGES_REQUESTED** [review]
The smaller concerns you will find in the code. But the general code structure is good, so I don't expect them to cause major problems.

My one big concern though is performance. On one of our larger customer databases there are about 1 million work packages and over 5 million journals of which the vast majority are associated to a work package. A query for historic data will always have to sort the journals (e.g.

```
...
WHERE "journals"."id" IN (
SELECT distinct on (journals.journable_type, journable_id) id FROM "journals" 
 WHERE "journals"."created_at" <= '2022-01-06 07:57:14.153220' AND journals.journable_type = 'WorkPackage'
 ORDER BY journals.journable_type ASC, "journals"."journable_id" ASC, "journals"."created_at" DESC)
 ```
 
 ) to find the most recent journal older than the time queried for. And an increase in journals will almost certainly increase the number of journals to sort assuming that most users will be less interested in comparing to a very old state. 

I already tried a couple of things: 

* Changing to an `EXISTS` which might perform better if the work package journals can be filtered by different conditions (e.g. `assigned_to_id = 5`) or to an inner join.
* Different ways of finding the most recent journal (e.g. "group by" and self left join)
* Denormalizing data by duplicating columns from `journals` and `work_packages` in the `work_package_journals` table.
* Adding indices, to help with the ordering.
* Adding a partial index only for `journable_type = 'WorkPackages' 
* Having a materialized view to precalculate the historic data structure containing `journals`, `work_packages` and `work_package_journals`

In the end, all my attempts performed poorly in the worst case which is not filtering on the `work_package_journals` table at all. Unfortunately, that worst case is not unlikely to take place. And in that case, a historic query tends to perform an order of magnitude worse than a similar non historic one.

What I would propose to do is try to identify the most likely timestamps for which users would like to get the baseline (e.g. 1 week ago, 1 month ago, ...). For those. we could create materialized views on a regular basis and they would just contain the journals necessary to know within that timeframe which journal is the most recent one. That would not mean that there would be only one journal per work package but there would be way fewer than the complete set of journals and could be sorted. A query could then switch to the appropriate view based on the time frame queried for.

A benefit of that approach is that the code in this PR could be simplified but we would have to pay for this by having to maintain the materialized views.

Maybe you have more ideas @fiedl and if so, I would be very interested in talking about them.

**Comment by fiedl** [comment]
## `updated_at` column

*Call 2022-12-21*

The `work_package_journals` table does not have an `updated_at` column as the `work_packages` table. When asking for the `updated_at` attribute, we need to map this to the `journals` table.

Originally, I have mapped the `updated_at` attribute to `journals.created_at`, assuming that each journal entry corresponds to a specific change. However, the journal entries can be updated when several changes are made in short succession in order to keep the activity log cleaner. 

- [x] Because journal entries are updated in this scenario, the last time of update is actually `journals.updated_at` rather than `journals.created_at`. ➝ https://github.com/opf/openproject/pull/11243/commits/d1962f14d441e8fb54f83c5bc4244c91badb0b5a

File: app/models/journable/historic_active_record_relation.rb (hunk 2 above)
Comment (fiedl) [reply app/models/journable/historic_active_record_relation.rb]: fixed in https://github.com/opf/openproject/pull/11243/commits/d1962f14d441e8fb54f83c5bc4244c91badb0b5a

File: app/models/journable/historic_active_record_relation.rb (hunk 4 above)
Comment (fiedl) [reply app/models/journable/historic_active_record_relation.rb]: `updated_at`: fixed in fixed in https://github.com/opf/openproject/pull/11243/commits/d1962f14d441e8fb54f83c5bc4244c91badb0b5a

File: app/models/journable/historic_active_record_relation.rb (hunk 10 above)
Comment (fiedl) [reply app/models/journable/historic_active_record_relation.rb]: fixed in https://github.com/opf/openproject/pull/11243/commits/d1962f14d441e8fb54f83c5bc4244c91badb0b5a

File: app/models/journable/timestamps.rb (hunk 6 above)
Comment (fiedl) [reply app/models/journable/timestamps.rb:99]: Possibly related: https://github.com/opf/openproject/commit/e9c6bce775a3bbe53d7a2560a89232172114ee5b

With this change, the columns missing in the journal data are returned as `null`, which prevents errors, but does not provide data for these columns, yet.

@ulferts Could you explain what the `version` and the `lock_version` attributes are, and how they are related?

Do all journables have a `version` column? Then, this mapping could be added here; otherwise, I could add the modification to a subclass.

Comment (ulferts) [reply app/models/journable/timestamps.rb:99]: @fiedl , the `version` column is part of the `journals` table. So every journable will have it as all are using the `journals` table. The `lock_version` on the other hand is `WorkPackage` specific. It is part of [AR::OptimisticLocking](https://api.rubyonrails.org/classes/ActiveRecord/Locking/Optimistic.html).

So regarding your question, adding a mapping here can not be generalized as only WorkPackage has the `lock_version` column. 

In the current implementation, where `null` is returned from the DB, that value is unfortunately mapped to `0` in the ruby world as that is the default value for that column. 

I don't care too much about that column's information it is just that it might lead to surprising behaviour. So if we could design some mechanism that clearly communicates to not trust that value that would be good enough for me.

File: app/models/journable/historic_active_record_relation.rb (hunk 7 above)
Comment (fiedl) [reply app/models/journable/historic_active_record_relation.rb:97]: 👍 resolved in https://github.com/opf/openproject/pull/11243/commits/41470cac7faebde554ebe98c7a1539adffef2bc9

**Comment by fiedl** [comment]
## Table alias for "journals" table

As suggested in https://github.com/opf/openproject/pull/11243#discussion_r1063195259, using a local table alias for the "journals" table would prevent accidental collisions when constructing a query like this:

```ruby
WorkPackage.at_timestamp(timestamp).includes(:journals).where(journals: {...})
```

However, introducing the alias is not trivial, I think:

```ruby
class Journable::HistoricActiveRecordRelation < ActiveRecord::Relation
  JOURNALS_TABLE_ALIAS = "journals2".freeze

  def add_timestamp_condition(relation)
    journal_relation = Journal.at_timestamp(timestamp)

    # TODO: journal_relation = journal_relation.with_table_alias(JOURNALS_TABLE_ALIAS)

    relation \
        .joins("INNER JOIN \"journals\" AS \"#{JOURNALS_TABLE_ALIAS}\"
            ON \"#{JOURNALS_TABLE_ALIAS}\".\"data_type\" = '#{model.journal_class.name}'
            AND \"#{JOURNALS_TABLE_ALIAS}\".\"data_id\" = \"#{model.journal_class.table_name}\".\"id\"".gsub("\n", "")) \
        .merge(journal_relation)
  end
end
```

Introducing the table alias in the sql literal above is no problem. But introducing the alias into the arel parts of the query has unintended side effects.

For example, when trying

```ruby
journal_relation.arel.source.left.table_alias = JOURNALS_TABLE_ALIAS
```

then the table alias is set not only for this relation, but on a deeper object within, i.e. `Journal.arel_table` itself. The above is equivalent to setting directly:

```ruby
Journal.arel_table.table_alias = JOURNALS_TABLE_ALIAS
```

This, unfortunately, has side effects on sql queries by `ActsAsJournalized` as well, e.g. `./lib_static/plugins/acts_as_journalized/lib/acts/journalized/reversion.rb:79`, which then would fail unless properly patched.

In conclusion:

- [ ] Either we find a way to set the table alias only for a single active-record relation,
- [x] or we postpone the attempt until we really need the table alias.

File: app/models/journable/historic_active_record_relation.rb (hunk 1 above)
Comment (fiedl) [reply app/models/journable/historic_active_record_relation.rb]: related discussion: https://github.com/opf/openproject/pull/11243#issuecomment-1421456967

File: app/models/journable/historic_active_record_relation.rb (hunk 8 above)
Comment (fiedl) [reply app/models/journable/historic_active_record_relation.rb]: *Phone call 2023-02-02*:

- Retrieving the historic value for the `position` column is not possible because this column is not journalized.
- We initialize the columns that are missing in the journalized table as `NULL` so that no database error is raised
- When trying to `pluck` or to retrieve such a value, we log a warning ➝ https://github.com/opf/openproject/pull/11243/commits/41470cac7faebde554ebe98c7a1539adffef2bc9, https://github.com/opf/openproject/pull/11243/commits/fe8ff8a1c12cf42bcebbda91d801808fb5c47036

File: app/models/journable/historic_active_record_relation.rb (hunk 9 above)
Comment (fiedl) [reply app/models/journable/historic_active_record_relation.rb]: Good catch, thanks! ➝ fixed in https://github.com/opf/openproject/pull/11243/commits/b838c6da17839c07a4ce3a82371a314119dc49ab

File: app/models/journable/timestamps.rb (hunk 6 above)
Comment (fiedl) [reply app/models/journable/timestamps.rb:99]: *Phone call 2023-02-02*:

- Retrieving the historic value for the `lock_version` column is not possible because this column is not journalized.
- We initialize the columns that are missing in the journalized table as `NULL` so that no database error is raised
- When trying to `pluck` or to retrieve such a value, we log a warning ➝ https://github.com/opf/openproject/pull/11243/commits/41470cac7faebde554ebe98c7a1539adffef2bc9, https://github.com/opf/openproject/pull/11243/commits/fe8ff8a1c12cf42bcebbda91d801808fb5c47036

File: app/models/journable/historic_active_record_relation.rb (hunk 11 above)
Comment (fiedl) [reply app/models/journable/historic_active_record_relation.rb:313]: ➝ https://github.com/opf/openproject/pull/11243#discussion_r1099272555

File: app/models/journal/timestamps.rb (hunk 3 above)
Comment (fiedl) [reply app/models/journal/timestamps.rb:72]: https://github.com/opf/openproject/pull/11243/commits/acf0d5580dbc3dfd0a5aa4b46f1e0f71401274dd

**Review by ulferts — APPROVED** [review]
@fiedl I am quite happy about this by now. Looks good and I'll merge it.

There is one thing that is not problematic right away but might lead to surprising behaviour later on and that lies in a call like 

```
puts WorkPackage.at_timestamp(5.days.ago).select(:id).to_sql

=> SELECT "work_package_journals"."id" FROM "work_package_journals" INNER JOIN "journals" ON "journals"."data_type" = 'Journal::WorkPackageJournal' AND "journals"."data_id" = "work_package_journals"."id" INNER JOIN (SELECT id, created_at FROM "work_packages") AS journables ON "journables"."id" = "journals"."journable_id" WHERE "journals"."journable_type" = 'WorkPackage' AND "journals"."id" IN (SELECT distinct on (journable_type, journable_id) id FROM "journals" WHERE "journals"."journable_type" = 'WorkPackage' AND "journals"."created_at" <= '2023-02-10 15:27:31.301822' ORDER BY "journals"."journable_type" ASC, "journals"."journable_id" ASC, "journals"."created_at" DESC)

```

In such a statement, the projection will be `work_package_journals.id`. If such a statement were to be used in a subselect, the wrong IDs would be referenced.

Compare that to the way `pluck` is handled explicitly. 

```
SELECT "journals"."journable_id" FROM ...
```

Ideally, the same mapping should be applied to `select` as it is for `pluck`. 

Nevertheless, this is not an immediate problem which is why I am merging this PR now.
